kafka/handle.go

```diff
@@ -28,9 +28,50 @@ import (
 */
 import "C"
 
+// OAuthBearerToken represents the data to be transmitted
+// to a broker during SASL/OAUTHBEARER authentication.
+type OAuthBearerToken struct {
+	// Token value, often (but not necessarily) a JWS compact serialization
+	// as per https://tools.ietf.org/html/rfc7515#section-3.1; it must meet
+	// the regular expression for a SASL/OAUTHBEARER value defined at
+	// https://tools.ietf.org/html/rfc7628#section-3.1
+	TokenValue string
+	// Metadata about the token indicating when it expires
+	// in terms of the number of milliseconds since the epoch; it must
+	// represent a time in the future
+	LifetimeMilliseconds int64
+	// Metadata about the token indicating the Kafka principal name
+	// to which it applies (for example, "admin")
+	Principal string
+	// SASL extensions, if any, to be communicated to the broker during
+	// authentication (all keys and values of which must meet the regular
+	// expressions defined at https://tools.ietf.org/html/rfc7628#section-3.1,
+	// and it must not contain the reserved "auth" key)
+	Extensions map[string]string
+}
+
 // Handle represents a generic client handle containing common parts for
 // both Producer and Consumer.
 type Handle interface {
+	// SetOAuthBearerToken sets the the data to be transmitted
+	// to a broker during SASL/OAUTHBEARER authentication. It will return nil
+	// on success, otherwise an error if:
+	// 1) the token data is invalid (meaning an expiration time in the past
+	// or either a token value or an extension key or value that does not meet
+	// the regular expression requirements as per
+	// https://tools.ietf.org/html/rfc7628#section-3.1);
+	// 2) SASL/OAUTHBEARER is not supported by the underlying librdkafka build;
+	// 3) SASL/OAUTHBEARER is supported but is not configured as the client's
+	// authentication mechanism.
+	SetOAuthBearerToken(oauthBearerToken OAuthBearerToken) error
```

Comment:
Since these are exposed on the Consumer, Producer, Admin.. clients, do these methods really need to be public on the Handle?

Reply:
Good question.  If they are not public then the code cannot be generalized to work with any client instance, and it is a Do-Not-Repeat-Yourself issue.  I had originally had repeated code as per the review comments on March 14th by jamesd, and I added these methods to `Handle` to allow consolidation of the mostly-repeated code into a single method.